File: docs/assignments/HW04/deliverables/automation/scripts/render-pdfs.mjs
import { readFileSync } from "node:fs";
import path from "node:path";
import { pathToFileURL } from "node:url";
import { chromium } from "@playwright/test";
import { marked } from "marked";

const reportDir = path.resolve("..", "reports");
const documents = ["main_report.md", "ai_critique.md", "ai_audit_report.md"];

const stylesheet = `
  @page { size: A4; margin: 17mm 15mm 19mm; }
  :root { color: #172033; font-family: "Segoe UI", Arial, sans-serif; font-size: 10.2pt; }
  body { line-height: 1.52; margin: 0; }
  h1, h2, h3 { color: #12355b; break-after: avoid; }
  h1 { border-bottom: 3px solid #2f75b5; font-size: 24pt; margin: 0 0 18px; padding-bottom: 9px; }
  h2 { border-bottom: 1px solid #a9bdd2; font-size: 16pt; margin-top: 28px; padding-bottom: 4px; }
  h3 { font-size: 12.5pt; margin-top: 21px; }
  p, li { orphans: 3; widows: 3; }
  table { border-collapse: collapse; font-size: 8.1pt; margin: 11px 0 18px; width: 100%; }
  th, td { border: 1px solid #aeb9c6; padding: 5px 6px; text-align: left; vertical-align: top; }
  th { background: #dce8f4; color: #12355b; font-weight: 650; }
  tr { break-inside: avoid; }
  code { background: #eef2f6; border-radius: 3px; font-family: Consolas, monospace; font-size: 8.7pt; padding: 1px 3px; }
  pre { background: #172033; color: #f4f7fa; overflow-wrap: anywhere; padding: 10px; white-space: pre-wrap; }
  pre code { background: transparent; color: inherit; padding: 0; }
  blockquote { border-left: 4px solid #2f75b5; color: #34485d; margin-left: 0; padding-left: 12px; }
  a { color: #165f9f; text-decoration: none; }
  strong { color: #102d4d; }
`;

const browser = await chromium.launch({ headless: true });
try {
  for (const fileName of documents) {
    const markdownPath = path.join(reportDir, fileName);
    const pdfPath = markdownPath.replace(/\.md$/i, ".pdf");
    const markdown = readFileSync(markdownPath, "utf8");
    const title = markdown.match(/^#\s+(.+)$/m)?.[1] ?? fileName;
    const body = await marked.parse(markdown, { gfm: true });
    const baseHref = pathToFileURL(`${reportDir}${path.sep}`).href;
    const page = await browser.newPage();
    await page.setContent(
      `<!doctype html><html lang="en"><head><meta charset="utf-8"><base href="${baseHref}"><title>${title}</title><style>${stylesheet}</style></head><body>${body}</body></html>`,
      { waitUntil: "load" },
    );
    await page.pdf({
      path: pdfPath,
      format: "A4",
      printBackground: true,
      displayHeaderFooter: true,
      headerTemplate: "<span></span>",
      footerTemplate: `<div style="font:8px 'Segoe UI',Arial;color:#607080;width:100%;padding:0 15mm;display:flex;justify-content:space-between"><span>23127404 · HW04 Automation Testing</span><span><span class="pageNumber"></span> / <span class="totalPages"></span></span></div>`,
      margin: { top: "17mm", right: "15mm", bottom: "19mm", left: "15mm" },
    });
    await page.close();
    console.log(`Rendered ${path.relative(process.cwd(), pdfPath)}`);
  }
} finally {
  await browser.close();
}

File: docs/assignments/HW06/specs/render_pdfs.py
import os
import markdown
from playwright.sync_api import sync_playwright

DOCS = [
    (
        "docs/assignments/HW06/deliverables/report/23127404_HW06_API_Testing_Report.md",
        "docs/assignments/HW06/deliverables/report/23127404_HW06_API_Testing_Report.pdf"
    ),
    (
        "docs/assignments/HW06/deliverables/ai/23127404_HW06_AI_Critique.md",
        "docs/assignments/HW06/deliverables/ai/23127404_HW06_AI_Critique.pdf"
    ),
    (
        "docs/assignments/HW06/deliverables/ai/23127404_HW06_AI_Audit_Report.md",
        "docs/assignments/HW06/deliverables/ai/23127404_HW06_AI_Audit_Report.pdf"
    ),
    (
        "docs/assignments/HW06/deliverables/cicd/23127404_HW06_CICD_Report.md",
        "docs/assignments/HW06/deliverables/cicd/23127404_HW06_CICD_Report.pdf"
    ),
]

CSS_STYLE = """
body {
    font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, "Helvetica Neue", Arial, sans-serif;
    line-height: 1.6;
    color: #24292e;
    padding: 20px;
    font-size: 13px;
}
h1 { color: #0366d6; font-size: 24px; border-bottom: 2px solid #eaecef; padding-bottom: 8px; margin-top: 0; }
h2 { color: #0366d6; font-size: 18px; border-bottom: 1px solid #eaecef; padding-bottom: 6px; margin-top: 24px; }
h3 { color: #1f2328; font-size: 15px; margin-top: 16px; }
table { border-collapse: collapse; width: 100%; margin: 15px 0; font-size: 12px; }
th, td { border: 1px solid #dfe2e5; padding: 7px 12px; text-align: left; }
th { background-color: #f6f8fa; font-weight: 600; color: #24292e; }
tr:nth-child(even) { background-color: #fcfcfc; }
code { background-color: #f6f8fa; padding: 2px 5px; border-radius: 4px; font-family: "SFMono-Regular", Consolas, "Liberation Mono", Menlo, monospace; font-size: 11.5px; color: #cf222e; }
pre { background-color: #f6f8fa; padding: 12px; border-radius: 6px; overflow-x: auto; border: 1px solid #e1e4e8; }
pre code { background-color: transparent; padding: 0; color: #24292e; }
blockquote { border-left: 4px solid #0969da; padding: 4px 15px; color: #57606a; margin: 15px 0; background-color: #f6f8fa; border-radius: 0 4px 4px 0; }
hr { border: 0; height: 1px; background: #e1e4e8; margin: 20px 0; }
"""

def render_all():
    with sync_playwright() as p:
        browser = p.chromium.launch()
        for md_path, pdf_path in DOCS:
            if not os.path.exists(md_path):
                print(f"Skipping missing file: {md_path}")
                continue
            with open(md_path, "r", encoding="utf-8") as f:
                md_text = f.read()
            html_body = markdown.markdown(md_text, extensions=["tables", "fenced_code", "nl2br"])
            full_html = f"""<!DOCTYPE html>
<html>
<head>
<meta charset="utf-8">
<style>{CSS_STYLE}</style>
</head>
<body>{html_body}</body>
</html>"""
            page = browser.new_page()
            page.set_content(full_html)
            page.pdf(
                path=pdf_path,
                format="A4",
                margin={"top": "18mm", "bottom": "18mm", "left": "15mm", "right": "15mm"},
                print_background=True
            )
            page.close()
            print(f"[OK] Rendered {os.path.basename(pdf_path)} ({os.path.getsize(pdf_path):,} bytes)")
        browser.close()

if __name__ == "__main__":
    render_all()
